btree/btree.c

```diff
@@ -3443,6 +3447,92 @@ static bool index_is_valid(struct level *lev)
 	return (lev->l_idx >= 0) && (lev->l_idx < node_count(lev->l_node));
 }
 
+/**
+ * btree_open_tick function is used to traverse through different states to
+ * facilitate the working of m0_btree_open().
+ *
+ * @param smop     represents the state machine operation
+ * @return int64_t returns the next state to be executed.
+ */
+int64_t btree_open_tick(struct m0_sm_op *smop)
+{
+	struct m0_btree_op    *bop  = M0_AMB(bop, smop, bo_op);
+	struct td             *tree = NULL;
+	int                    i    = 0;
+	struct m0_btree_oimpl *oi   = bop->bo_i;
+	struct segaddr         curr_addr;
+
+	switch(bop->bo_op.o_sm.sm_state)
+	{
+	case P_INIT:
+		if (!addr_is_aligned(bop->b_data.addr)) {
+			bop->bo_op.o_sm.sm_rc = M0_ERR(-EFAULT);
+			return P_DONE;
+		} else {
+
+			curr_addr = segaddr_build(bop->b_data.addr,
+						  calc_shift(
+						  bop->b_data.num_bytes));
+
+			m0_rwlock_write_lock(&trees_lock);
+			M0_ASSERT(trees_loaded <= ARRAY_SIZE(trees));
+
+			/**
+			 *  If existing allocated tree is found then return it
+			 *  after increasing the reference count and capturing
+			 *  the btree data.
+			 */
+			if (trees_loaded) {
+				for (i = 0; i < ARRAY_SIZE(trees); i++) {
+					tree = &trees[i];
+					m0_rwlock_write_lock(&tree->t_lock);
+					if (tree->r_ref != 0) {
+						M0_ASSERT(tree->t_root != NULL);
+						if (tree->t_root->n_addr.as_core
+						== curr_addr.as_core) {
+
+							tree->r_ref++;
+							bop->b_data.tree->t_desc
+							= tree;
+							bop->b_data.tree->t_lock
+							= tree->t_lock;
+							bop->b_data.tree->
+							t_height = tree->
+								   t_height;
+							bop->b_data.tree->t_type
+							= tree->t_type;
+
+							m0_rwlock_write_unlock(
+							&tree->t_lock);
+							m0_rwlock_write_unlock(
+							&trees_lock);
+							return P_DONE;
+						}
+					}
+					m0_rwlock_write_unlock(&tree->t_lock);
```

Comment:
If you are not going inside the line 3489 `if check`, then you are not releasing `trees_lock` lock.

Reply:
With tree_get(), this part wont be needed.
So it will be removed in the upcoming commit.

Also, when the said tree_descriptor is freed using `tree_put()` it is the one which will release the trees_lock.